src/libstd/sync/remutex.rs

```diff
@@ -0,0 +1,278 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+#![unstable(feature = "reentrant_mutex", reason = "new API")]
+
+use prelude::v1::*;
+
+use cell::{RefCell, Ref, RefMut};
+use borrow::{Borrow, BorrowMut};
+use marker;
+use sync::poison::{self, TryLockError, TryLockResult, LockResult};
+use sys::mutex as sys;
+use fmt;
+
+/// A re-entrant mutual exclusion
+///
+/// This mutex will block *other* threads waiting for the lock to become available. The thread
+/// which has already locked the mutex can lock it multiple times without blocking, preventing a
+/// common source of deadlocks.
+///
+/// # Examples
+///
+/// ```
+/// use std::sync::ReentrantMutex;
+///
+/// fn locking_mod(data: ReentrantMutex<i32>) {
+///     let mut data = data.lock().unwrap();
+///     *data += 1;
+/// }
+///
+/// let data = ReentrantMutex::new(0);
+/// {
+/// // We take the lock twice here. It won’t deadlock.
+/// let lock = data.lock();
+/// locking_mod(data);
+/// }
+/// ```
+pub struct ReentrantMutex<T> {
+    inner: Box<sys::ReentrantMutex>,
+    poison: poison::Flag,
+    data: RefCell<T>,
```

Comment:
There are other forms of interior mutability, however, such as `Cell` or `UnsafeCell` (rolling your own). Additionally it may be the case that the top-level doesn't actually need to be protected with a `RefCell` but only some inner type (e.g. one or two fields).